packages/panzer/disc-fe/src/evaluators/Panzer_GatherSolution_Epetra_impl.hpp

```diff
@@ -210,38 +215,58 @@ template<typename TRAITS,typename LO,typename GO>
-   //       The "getGIDFieldOffsets may be expensive.  However the
-   //       "getElementGIDs" can be cheaper. However the lookup for LIDs
-   //       may be more expensive!
-
-   Epetra_Vector & x = *x_;
-
-   // gather operation for each cell in workset
-   for(std::size_t worksetCellIndex=0;worksetCellIndex<localCellIds.size();++worksetCellIndex) {
-      std::size_t cellLocalId = localCellIds[worksetCellIndex];
-
-      LIDs = globalIndexer_->getElementLIDs(cellLocalId);
-
-      // loop over the fields to be gathered
-      for (std::size_t fieldIndex=0; fieldIndex<gatherFields_.size();fieldIndex++) {
-         int fieldNum = fieldIds_[fieldIndex];
-         const std::vector<int> & elmtOffset = globalIndexer_->getGIDFieldOffsets(blockId,fieldNum);
+  std::vector<int> LIDs;
+
+  // For convenience, pull out some objects from the workset.
+  std::string blockId = this->wda(workset).block_id;
+  const std::vector<std::size_t>& localCellIds =
+    this->wda(workset).cell_local_ids;
+
+  // NOTE: A reordering of these loops will likely improve performance
+  //       The "getGIDFieldOffsets may be expensive.  However the
+  //       "getElementGIDs" can be cheaper. However the lookup for LIDs
+  //       may be more expensive!
+
+  // Gather operation for each cell in the workset.
+  int numOwnedIndices = globalIndexer_->getNumOwned();
+  for (std::size_t worksetCellIndex = 0;
+    worksetCellIndex < localCellIds.size(); ++worksetCellIndex)
+  {
+    std::size_t cellLocalId = localCellIds[worksetCellIndex];
+    LIDs = globalIndexer_->getElementLIDs(cellLocalId);
 
-         // loop over basis functions and fill the fields
-         for(std::size_t basis=0;basis<elmtOffset.size();basis++) {
-            int offset = elmtOffset[basis];
-            int lid = LIDs[offset];
-            (gatherFields_[fieldIndex])(worksetCellIndex,basis) = x[lid];
-         }
+    // Loop over the fields to be gathered.
+    for (std::size_t fieldIndex = 0; fieldIndex < gatherFields_.size();
+      ++fieldIndex)
+    {
+      int fieldNum = fieldIds_[fieldIndex];
+      const std::vector<int>& elmtOffset =
+        globalIndexer_->getGIDFieldOffsets(blockId, fieldNum);
+
+      // Loop over the basis functions and fill the fields.
+      for (std::size_t basis = 0; basis < elmtOffset.size(); ++basis)
+      {
+        int offset = elmtOffset[basis];
+        int lid    = LIDs[offset];
+        if (x_.is_null())
+        {
+          if (lid < numOwnedIndices)
+          {
+            Teuchos::RCP<const Epetra_Vector> xOwned =
+              Thyra::get_Epetra_Vector(*ownedMap_, xOwned_);
+            (gatherFields_[fieldIndex])(worksetCellIndex, basis) =
+              (*xOwned)[lid];
+          }
+          else
+            (gatherFields_[fieldIndex])(worksetCellIndex, basis) =
+              (*xGhosted_)[lid - numOwnedIndices];
+        }
+        else
+          (gatherFields_[fieldIndex])(worksetCellIndex, basis) =
+            (*x_)[lid];
```

Comment:
The `if (x_.is_null())` can be read `if (we're doing this the new way)`.  Instead of using the local ID to index into `x_`, it will now index into either `xOwned_` or `xGhosted_`.  The local IDs are ordered such that all owned indices come first, followed by the ghosted ones.  For the time being, the old way with `x_` is still supported until we're able to verify this new way works.